### PR TITLE
[5.5] Use table aliases when calling a self-referencing HasManyThrough

### PR DESCRIPTION
**Pull request by JayBizzle**
See #21259 for full-details.

This fix has been implemented in a similar way that table aliases are applied when using a self-referencing `BelongsTo` relationship.

I have targeted this at 5.5, but not sure if this is classed as a breaking change as some people may be relying on the current broken behaviour. If so, I can target this at 5.6 no problem.

### CODE REVIEW

**Comment by jmarcher** [comment]
Can you please include some tests?

**Comment by JayBizzle** [comment]
Based on the functionality of the `BelongsTo` relation, there are no tests

**Comment by themsaid** [comment]
@JayBizzle you can write an integration test where you hit an actual DB and make sure it returns results.

**Comment by JayBizzle** [comment]
Yes, an integration test should be no problem

**Comment by themsaid** [comment]
@JayBizzle I think the example you have is wrong, use case doesn't feel right to me, can you explain in plain english the relationship between the user and his referrals?

**Comment by themsaid** [comment]
Here's an example self-referencing HasManyThrough use case:

```
<?php

namespace Illuminate\Tests\Integration\Database\EloquentHasManyThroughTest;

use Orchestra\Testbench\TestCase;
use Illuminate\Support\Facades\Schema;
use Illuminate\Database\Eloquent\Model;

/**
 * @group integration
 */
class EloquentHasManyThroughTest extends TestCase
{
    protected function getEnvironmentSetUp($app)
    {
        $app['config']->set('app.debug', 'true');

        $app['config']->set('database.default', 'testbench');

        $app['config']->set('database.connections.testbench', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }

    public function setUp()
    {
        parent::setUp();

        Schema::create('users', function ($table) {
            $table->increments('id');
            $table->integer('team_id')->nullable();
            $table->string('name');
        });

        Schema::create('teams', function ($table) {
            $table->increments('id');
            $table->integer('owner_id');
        });
    }

    /**
     * @test
     */
    public function basic_create_and_retrieve()
    {
        $user = User::create(['name' => str_random()]);

        $team1 = Team::create(['owner_id' => $user->id]);
        $team2 = Team::create(['owner_id' => $user->id]);

        $member1 = User::create(['name' => str_random(), 'team_id' => $team1->id]);
        $member2 = User::create(['name' => str_random(), 'team_id' => $team1->id]);

        $notMember = User::create(['name' => str_random()]);


        // Returns the correct results for a self referencing relationship
        $user->members->toArray();
    }
}


class User extends Model
{
    public $table = 'users';
    public $timestamps = false;
    protected $guarded = ['id'];

    public function members()
    {
        return $this->hasManyThrough(self::class, Team::class, 'owner_id', 'team_id');
    }
}

class Team extends Model
{
    public $table = 'teams';
    public $timestamps = false;
    protected $guarded = ['id'];
}
``

It works fine, so I don't think there's a bug that needs to be fixed.

**Comment by JayBizzle** [comment]
Give me a second to go through your example....

**Comment by themsaid** [comment]
@JayBizzle I've updated the example with better relationship name for clarity.

**Comment by JayBizzle** [comment]
Okay, testing now....

**Comment by JayBizzle** [comment]
@themsaid In your example, if you did the following...

```
User::has('teamMates')->get();
```

You would get no results, whereas you should get 1 which you do after applying my fix.

Your example isn't quite what we were trying to achieve, but it still suffers the same problem.

**Comment by themsaid** [comment]
@JayBizzle I see, can you rebase your PR with my tests instead of the those your wrote? We can start checking the approach you took and see if there's a simpler one.

**Comment by JayBizzle** [comment]
Yes, ill get that done for you.

I guess this PR name is a bit misleading as this issue doesn't affect ALL self-referencing HasManyThrough relationship calls, only ones that use sub-queries.

**Comment by themsaid** [comment]
@JayBizzle yeah it got me confused, but it's ok now that we can see the issue.

**Comment by themsaid** [comment]
@JayBizzle git went crazy, I suggest that you open a new PR with just the changes in https://raw.githubusercontent.com/JayBizzle/framework/54edc2d20e96e2f429d24c21e33d6a880cd5d044/src/Illuminate/Database/Eloquent/Relations/HasManyThrough.php, you can copy the file in your new PR and that's it.

**Comment by JayBizzle** [comment]
Yeah, i totally messed that up :)
